servicetalk-concurrent-api/src/testFixtures/java/io/servicetalk/concurrent/api/TestCollectingPublisherSubscriber.java

```diff
@@ -0,0 +1,213 @@
+ *     sub.awaitOnComplete(false); // wait for the terminal signal, ignore if there are unconsumed onNext signals.
+ *     List&lt;String&gt; onNextSignals = sub.pollAllOnNext();
+ *     // verify all onNextSignals occurred in the expected order
+ * </pre>
+ * @param <T> The type of data in {@link #onNext(Object)}.
+ */
+public final class TestCollectingPublisherSubscriber<T> implements Subscriber<T> {
+    private static final Object NULL_ON_NEXT = new Object();
+    private static final Throwable ON_COMPLETE = unknownStackTrace(new RuntimeException("onComplete"),
+            TestCollectingPublisherSubscriber.class, "onComplete");
+    private final BlockingQueue<Object> items = new LinkedBlockingQueue<>();
+    private final CountDownLatch onTerminalLatch = new CountDownLatch(1);
+    private final CountDownLatch onSubscribeLatch = new CountDownLatch(1);
+    @Nullable
+    private Throwable onTerminal;
+    @Nullable
+    private Subscription subscription;
+
+    @Override
+    public void onSubscribe(final Subscription subscription) {
+        this.subscription = subscription;
+        onSubscribeLatch.countDown();
+    }
+
+    @Override
+    public void onNext(@Nullable final T t) {
+        items.add(t == null ? NULL_ON_NEXT : t);
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+        onTerminal = t;
+        onTerminalLatch.countDown();
+    }
+
+    @Override
+    public void onComplete() {
+        onTerminal = ON_COMPLETE;
+        onTerminalLatch.countDown();
+    }
+
+    /**
+     * Block until {@link #onSubscribe(Subscription)}.
+     *
+     * @return The {@link Subscription} from {@link #onSubscribe(Subscription)}.
+     * @throws InterruptedException if an interrupt occurs while blocking for waiting for
+     * {@link #onSubscribe(Subscription)}.
+     */
+    public Subscription awaitSubscription() throws InterruptedException {
+        onSubscribeLatch.await();
+        if (subscription == null) {
+            throw new IllegalStateException("subscription is null");
+        }
+        return subscription;
+    }
+
+    /**
+     * Blocks until the next {@link #onNext(Object)} method invocation.
+     *
+     * @return item delivered to {@link #onNext(Object)}.
+     * @throws InterruptedException if an interrupt occurs while blocking for the next item.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public T takeOnNext() throws InterruptedException {
+        Object item = items.take();
+        return item == NULL_ON_NEXT ? null : (T) item;
+    }
+
+    /**
+     * Consume all currently available {@link #onNext(Object)} signals.
+     *
+     * @return {@link List} containing all currently available {@link #onNext(Object)} signals.
+     */
+    @SuppressWarnings("unchecked")
+    public List<T> pollAllOnNext() {
+        List<T> consumedItems = new ArrayList<>();
+        Object item;
+        while ((item = items.poll()) != null) {
```

Comment:
Consider using `drainTo()` and `stream()` for a more succinct representation as we can absorb higher allocations in test code:

```java
List<Object> consumedItems = new ArrayList<>();
items.drainTo(consumedItems);
return consumedItems.stream().map(item -> item == NULL_ON_NEXT ? null : (T) item).collect(toList());
```

Reply:
done